Add a reload command

RhythmicSys opened the pull request:
- Adds reload command
- Adds permission for reload command
- Adds method to save and clear current runnables
- Updates the build java version to java 17 (couldn't build without it)
- Also adjusts the config loader so new values are automatically added to configs (does not add comments)

RhythmicSys committed:
Add reload command
- Adds reload command
- Adds permission for reload command
- Adds method to save and clear current runnables
- Updates the build java version to java 17 (couldn't build without it)

syldium commented on spigot/src/main/java/com/github/syldium/nethertree/commands/ReloadCommand.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Override
+    public boolean onCommand(@NotNull CommandSender sender, @NotNull Command command, @NotNull String s, @NotNull String[] args) {
+        if (!sender.hasPermission(NetherTreePermissions.NETHER_TREE_ADMIN)) {
+            sender.sendMessage(ChatColor.RED + "You don't have permission to use this command.");
+            return true;
+        }
```
```suggestion
```
Permissions are already defined in the `plugin.yml` file.

syldium commented on spigot/src/main/java/com/github/syldium/nethertree/commands/ReloadCommand.java:
```diff
@@ -0,0 +1,35 @@
+            return true;
+        }
+        plugin.reloadConfig();
+        if (plugin.getRunnablesManager() != null) {
```
The `RunnableManager` should never be null: it's a registry of runnables that may be empty but not null.

syldium commented on spigot/src/main/java/com/github/syldium/nethertree/runnable/RunnablesManager.java:
```diff
@@ -119,6 +119,21 @@ public void save() {
     }
 
+    public void saveCancelAndClear() {
+        getFileConfiguration(false).ifPresent(config -> {
+            for (Map.Entry<UUID, DecayRunnable> entry : this.runnables.entrySet()) {
+                config.set(entry.getKey().toString(), entry.getValue().serialize());
+                entry.getValue().cancel();
+            }
+            this.runnables.clear();
+            try {
+                config.save(this.fileConfig);
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        });
```
```suggestion
        save();
        this.runnables.clear();
```
I think the `RunnablesManager` class may just be re-instantiated in case you want to reload.

RhythmicSys replied:
lol, oops, yep

syldium commented on spigot/src/main/java/com/github/syldium/nethertree/NetherTreePlugin.java:
```diff
@@ -52,6 +53,14 @@ public DecayRunnable getRunnable(World world) {
         return this.runnablesManager.getRunnable(world);
     }
 
+    public RunnablesManager getRunnablesManager(){
+        return this.runnablesManager;
+    }
+
+    public void setRunnablesManager(RunnablesManager runnablesManager){
+        this.runnablesManager = runnablesManager;
+    }
+
```
`RunnablesManager` is an implementation detail of the plugin. If you just want to expose a way to reload the plugin, the `JavaPlugin#reloadConfig` method can be overridden to re-read the files and re-create the appropriate classes:
```java
@Override
public void reloadConfig() {
    // Save runnables since they will be recreated
    this.runnablesManager.save();
    // Re-read the config.yml file: the default behavior of the JavaPlugin class
    super.reloadConfig();
    // Re-create our classes
    this.loadConfig();
}
```

RhythmicSys replied:
Attempted to do it this way in the updated version, though I wasn't sure how to do it without reorganizing the declarations a bit in the main class.

syldium commented on spigot/pom.xml:
```diff
@@ -13,7 +13,7 @@
     <description>Decays the leaves of the Fungi trees (Nether trees)</description>
     <properties>
-        <java.version>1.8</java.version>
+        <java.version>17</java.version>
```
```suggestion
        <java.version>1.8</java.version>
```
You are welcome to build with a more recent version of Java, but at the end this plugin still targets Minecraft 1.16, which requires only Java 8*.

_*and given the small scale of this plugin, it is still affordable to support an outdated Java version._

RhythmicSys replied:
I was unable to actually build it without changing the java version, it wouldn't compile. I had attempted to reset my IDE before doing that, somehow now, after changing it back to 1.8, it's working so I legitimately have no idea what was wrong there.

syldium reviewed (changes requested):
Thanks for your PR! I truly appreciate having a reload functionality in plugins. Please note that the `DropCalculator` instance is also configuration-dependent, so it will need to be recreated when the plugin is reloaded.

RhythmicSys committed:
Apply requested changes
- Revert java version to java 8
- Adjust reload setup to override the bukkit config reload
- Add Drop Calculator and Tree Handler to the classes reloaded

syldium committed:
Refactor reloadConfig

syldium committed:
Rename command parameter

syldium committed:
Make nethertree.admin.reload as a child permission

syldium reviewed (approved):
Thanks for your contribution!
